### PR TITLE
Fixes #24546 - add uniq constraint on subnet.name

### PR DESCRIPTION
**Pull request by iNecas**
Concurrent requests on subnet create can lead to duplicate records. This
is calling issues in production when bulk importing hosts from the same
subnet via user scripting.

- [x] - ~~merge taxonomies of the merged subnets~~ - don't merge subnets from different taxonomies
- [x] - ensure we don't merge subnets with different network, but ensure uniqueness there.

### CODE REVIEW

File: db/migrate/20180806151925_add_subnet_name_unique_constraint.rb
Comment (adamruzicka) [inline db/migrate/20180806151925_add_subnet_name_unique_constraint.rb]: How about just `#pluck`ing the ids here and doing something like `Subnet.unscoped.where(id: loosers).destroy_all` on the last line instead of retrieving the ids from the `Subnet` objects all the time?

Comment (iNecas) [reply db/migrate/20180806151925_add_subnet_name_unique_constraint.rb]: I went for `destroy` just in case there was any cleanup happening in the after_destroy hooks. On the other hand, I can imagine this causing troubles in case the after_destoy hooks used some attributes that were not migrated yet. I will update that later today

**Comment by lzap** [comment]
This change ignores taxonomy, I am not sure if we can do this on the database level.

**Comment by iNecas** [comment]
I will add code for merging taxonomies as well. I don't see any other way than doing it on database level: I want to avoid active record hooks because it risks casing the migrations to fail later. Another reason why I think it's safe is if the subnet had duplicate name, it would prevent the subnet to be saved later, so the user actually couldn't change the taxonomies of any of the duplicate subnets later.

**Comment by lzap** [comment]
But still user in Org B won't be able to create "my subnet" if it's already present in Org A. This is just a name which is not important, we use database ID for referencing.

But most importantly, after reading the case, I am not sure if name is the real problem. What's more important is the subnet network address. We often use method called `subnet_for` which is supposed to return correct subnet for given network and if there are multiple subnets created with different names but same network.

So to be sure to fix the issue with multiple subnets found by this call, I think we need a validator for network address. Taxonomy is still issue here. I think we should raise a thread on the list whether we should drop taxonomy from Subnets. Not sure about it.

**Comment by ares** [comment]
I don't think we should drop taxonomies for subnets. It totally makes sense from filtering point of view. Think of locations.

We know our multitenancy model is limited, we enforce various uniq check on many places, but a lot of users use taxonomies for RBAC or quick filtering.

**Comment by lzap** [comment]
The problem with Rails level validation is that it won't help with concurrency issues at all. And on the db level we need to ditch taxonomy. Not really sure what the correct approach is. Was CURL re-sending delayed requests? Because I don't understand why it happened in the first place.

**Comment by iNecas** [comment]
This PR is purely about enforcing the current unique constraint. There is not magic in the curl, pure race condition in Rails on checking the uniqueness and creating the record.

I agree with the taxonomies case: our taxonomies as they are written today, are not just covering the multi-tenancy and isolation on org-level well today and this PR doesn't try to solve that: it requires more design and time to fix, therefore I will not even follow the conversation in this PR about it to not derail it.

This PR is purely about bulk registration of clients from users subnet, where the registration script creates the subnet in case it's not present (that's what the reproducer script simulates), and leads to multiple records you can't use.

Adding the unique constraint on the subnet is yet another issue I don't want to address in this PR: adding a new uniq constraint into the model brings quite a lot of risk into the migration, therefore I would like to ask twice if the benefit of it is so bit.

File: db/migrate/20180806151925_add_subnet_name_unique_constraint.rb
Comment (ekohl) [inline db/migrate/20180806151925_add_subnet_name_unique_constraint.rb]: length

**Review by ekohl — CHANGES_REQUESTED** [review]
Does it mean some users can lose subnets or would they now see duplicated entries?

**Comment by iNecas** [comment]
No subnet would be lost, this is purely about reducing the duplicates that get created during the race-condition (there is no other way to get to this situation). The duplicate records mean that the user actually can't edit any of the subnets, as any attempt to save would actually mean validation failure.

**Comment by ekohl** [comment]
But you destroy a subnet. What if the name matches but not the actual network? Or a property like the reverse DNS proxy / DHCP proxy mismatch?

**Comment by iNecas** [comment]
Ok, I don't think there is high change we would be hitting this case, but what I could do in case the attributes don't match, to add suffix `-1, -2…` for those, that don't match. Would that be ok with others, before I go implementing that? Subnets that match in all arguments would still be reduced into one.

**Comment by ekohl** [comment]
@iNecas that sounds good.
Another thing that I'm not entirely sure about is whether a user could have access to an object via a location where after deduplication they can no longer access it. Not sure we can ignore taxonomies because of that.

**Review by lzap — CHANGES_REQUESTED** [review]
This patch does not make much sense and only makes things worse in regard to taxonomy. If you make name unique across all requests then what's much more important is subnet network address which can lead to possible issues, it's not the name.

I don't agree with fixing just name (which is IMHO not important we reference subnets by ID) and leaving much more sensitive field (network address, we search for networks via `for_subnet` call) as is.

This will only help in cases where name = network address. For users hitting the issue, we need to provide script to delete extra subnets. And then work on proper solution rather than delivering half-baked one.

**Comment by lzap** [comment]
> This PR is purely about enforcing the current unique constraint

And the current proposal is enforcing it on a wrong field, it's `network` column we need to enforce on, not name.

> Adding the unique constraint on the subnet is yet another issue I don't want to address in this PR: adding a new uniq constraint into the model brings quite a lot of risk into the migration, therefore I would like to ask twice if the benefit of it is so bit.

No, it is the very same issue and it should be fixed. There is no difference in risk - you are doing the very same migration already in this PR:

     winner, *loosers = Subnet.unscoped.where(name: name)

Tell me more about the risk please.

I don't want much, all I say (and please correct me if I am wrong) is we need to enforce on `network` column, that's the important one. If we ever gonna do such a migration, let's do it once and properly. Because the current proposal does not fix the *problem with multiple subnets*, it only fixes *problem of multiple subnets with same name*. Imagine you have two perfectly different subnets (by network address) but by mistake named with the same name - this migration will merge it creating data loss.

**Comment by iNecas** [comment]
@lzap I'm sorry but I don't agree - this are two distinct issues:

1. The issue I'm dealing with right now happens only with the race condition described in the issue. There is no other way to get into this situation. We currently rely on uniqueness of the name, and the reported issue leads to breaking this assumption, that we have in many places. As a consequence, the subnets are created as invalid in database, and you can't even update it from the application (not mentioning also the confusions it causes in the UI select boxes.

> And the current proposal is enforcing it on a wrong field

No. the enforcement is already there: it's just fixing that enforcement.

> Imagine you have two perfectly different subnets (by network address) but by mistake named with the same name - this migration will merge it creating data loss.

No, please read the issue description carefully again: there is very little chance this situation will happen. I've proposed https://github.com/theforeman/foreman/pull/5923#issuecomment-412544776 to address this risk.

2. The issue you are talking about is creating two subnets with different name, that would define the same subnet: there is much higher chance to get to this state: therefore the risk of the migration is much higher as well - this is not just about race-condition anymore. Also, this requires changes in the UI as well to not rely on unique name, but rather than on network.

3. There is also another issue with taxonomies, that is also independent of this one.

I don't agree this is half-baked solution. This is full solution of (1) that addresses a specific issue the user hit: they also followed a good practice of having the IP as part of the name, so they don't run into the issue . The user didn't hit (2) nor (3), which leads me to conclusion, that those are different issues.

I think 2. and 3. are much bigger in scope than this one, and I would like to at least make sure the enforcement we do currently is something we can rely on, until we have some different model.

**Comment by ares** [comment]
I think there was a bit of misunderstanding of what this PR is trying to solve. I like the desire to move subnets validations further and potentially at some point even make our taxonomies supporting full multi-tenancy, but I don't think it should happen as part of this PR. If you feel strongly that we need to improve this now, I'd recommend opening a separate issue and start working on the patch. IMHO this patch fixes the original issue. And there are other issues that we know about.

**Review by lzap — DISMISSED** [review]
I've worked with several customer environments which don't follow the rule of having subnet network address in the name column, yet they had meaningful names for them. Therefore this makes me think that the enforcement on name is incorrect, name does not matter.

You say that since we already have this in our codebase your patch is good, that's not the case. It's wrong from the day one and we should be deleting the validation from Rails code and adding new one on database level on subnet network. That will also prevent misbehavior you see with your reproducer.

Having said that, I agree the risk is higher that I thought, I initially missed that we already have the validation in Rails layer. Therefore I am fine with just moving this into database layer if you are planning backporting this, that's a reasonable solution.

**Comment by iNecas** [comment]
Thanks. I will still add there some logic to prevent merging two subnet objects with different network, and ensure we merge taxonomies on the objects as well. Will update once new changes are there.

**Comment by ekohl** [comment]
I don't think we can really make the subnet column unique because of NAT. In a multi-org or multi-location setup you can use the same RFC1918 ranges multiple times. It's not nice, but it can happen. A perfectly valid case can be two organizations that both have a `LAN` subnet with the same network but no overlap in users at all. Merging those can result in odd effects. If no proxy is assigned because they manually manage IPs, merging could result in duplicated IPs within a subnet (based on DB ID).

Would it be bad to never merge and always rename? Mention in the release notes this can happen and how admins can find duplicated subnets?

**Comment by iNecas** [comment]
I'm going to merge only if all the params are the same: that should be safe enough for mitigating the rece condition.

**Comment by ekohl** [comment]
What I meant was that DNS proxy `null` and DHCP proxy `null` on both could still be a difference due to taxonomies in that they don't share anything. Let's say Org A has a subnet `LAN` with `192.168.1.0/24` and Org B has a subnet `LAN` with `192.168.1.0/24`. They have no proxies attached so they are only administrative. That means `host-x.org-a.com` can have `192.168.1.100` and `host-y.org-b.com` can also have `192.168.1.100`. After merging they are referring to the same subnet so now the tuple `(ip, subnet)` for a host is no longer unique where it previously was. I realize that this might be far fetched and unlikely to occur, but could it happen?

**Comment by iNecas** [comment]
@ekohl IMO the probability that they would create this objects at the same time, with the same name, but in different orgs is next to zero, but I can check also the taxonomies, and instead of merging, just adding unique name, if you you would feel better after that.

**Comment by iNecas** [comment]
Ready for re-review: this version doesn't try merge the subnets until all the attributes + taxonomies match. If the don't, it will start adding `-1`, `-2`, `-3`… suffixes to ensure uniqueness after the migration.

File: db/migrate/20180806151925_add_subnet_name_unique_constraint.rb
Comment (lzap) [inline db/migrate/20180806151925_add_subnet_name_unique_constraint.rb]: Typo in lenght.

File: db/migrate/20180806151925_add_subnet_name_unique_constraint.rb
Comment (lzap) [inline db/migrate/20180806151925_add_subnet_name_unique_constraint.rb]: Use `say` or this could be candidate for `say_with_time`: https://rails-bestpractices.com/posts/2010/08/19/use-say-and-say_with_time-in-migrations-to-make-a-useful-migration-log/

File: db/migrate/20180806151925_add_subnet_name_unique_constraint.rb
Comment (lzap) [inline db/migrate/20180806151925_add_subnet_name_unique_constraint.rb]: Why you do this?

File: db/migrate/20180806151925_add_subnet_name_unique_constraint.rb
Comment (lzap) [inline db/migrate/20180806151925_add_subnet_name_unique_constraint.rb]: Uh, okay. But this is no concurrency, a hash table might serve as well.

File: db/migrate/20180806151925_add_subnet_name_unique_constraint.rb
Comment (lzap) [inline db/migrate/20180806151925_add_subnet_name_unique_constraint.rb]: I'd use `destroy_all!` or something like that - we want to error out early during this process if deletion does not go well.

**Review by lzap — CHANGES_REQUESTED** [review]
Few small things, but it's a lot of code. Can you write one or few tests? It should be pretty easy giving us much more confidence. This can bite.

File: db/migrate/20180806151925_add_subnet_name_unique_constraint.rb
Comment (iNecas) [reply db/migrate/20180806151925_add_subnet_name_unique_constraint.rb]: To not merge subnets that have different taxonomies

File: db/migrate/20180806151925_add_subnet_name_unique_constraint.rb
Comment (iNecas) [reply db/migrate/20180806151925_add_subnet_name_unique_constraint.rb]: Could you be more specific on different solution? The main reason for the suffix is we can't be sure 100% that there was not already a subnet present with `#{name}-#{new_suffix}`, so I'm not sure why we're talking about concurrency here?

File: db/migrate/20180806151925_add_subnet_name_unique_constraint.rb
Comment (iNecas) [reply db/migrate/20180806151925_add_subnet_name_unique_constraint.rb]: https://stackoverflow.com/questions/32928476/are-there-destroy-all-delete-all-in-rails

File: db/migrate/20180806151925_add_subnet_name_unique_constraint.rb
Comment (lzap) [reply db/migrate/20180806151925_add_subnet_name_unique_constraint.rb]: Okay, that's fair.

File: db/migrate/20180806151925_add_subnet_name_unique_constraint.rb
Comment (iNecas) [reply db/migrate/20180806151925_add_subnet_name_unique_constraint.rb]: fixed

File: db/migrate/20180806151925_add_subnet_name_unique_constraint.rb
Comment (iNecas) [reply db/migrate/20180806151925_add_subnet_name_unique_constraint.rb]: fixed

**Review by iNecas — COMMENTED** [review]
I believe I've addressed all the comments in the review

**Review by lzap — APPROVED** [review]
Nice, waiting for jenkins and then I'll test and merge.

**Comment by lzap** [comment]
Unless @ekohl has comments.

**Review by ekohl — APPROVED** [review]

**Comment by lzap** [comment]
Tests do not seem relevant, will test locally tomorrow:

```
Expected response to be a <2XX: success>, but was a <500: Internal Server Error>
Response body: {
  "error": {"message":"undefined method `class_name' for nil:NilClass\nDid you mean?  class_eval"}
}
 (Minitest::Assertion)
/var/lib/workspace/workspace/test_develop_pr_core@3/database/postgresql/ruby/2.4/slave/fast/test/controllers/api/v2/fact_values_controller_test.rb:64   
```

**Comment by ares** [comment]
@lzap any update on testing?

**Review by lzap — APPROVED** [review]

**Comment by lzap** [comment]
Thank you very much for your contribution, this is now merged.

**Pull request merged**
